### PR TITLE
Add ENS16x (air quality) and ENS210 (temp & RH) sensors

### PR DESCRIPTION
**Pull request by chrfriese123**
## Description:

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [x] The code change is tested and works with Tasmota core ESP32 V.2.0.11
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

**Comment by barbudor** [comment]
Please remove *.bak files

File: tasmota/include/tasmota_configurations.h
Comment (barbudor) [inline tasmota/include/tasmota_configurations.h:115]: Do not enable your drivers by default
Also comment the increase in code size for ESP8266 (and memory too would be great)
Comment those lines so people are aware of what should be used

File: BUILDS.md
Comment (barbudor) [inline BUILDS.md:136]: Do not add your drivers by default

File: tasmota/include/tasmota_configurations_ESP32.h
Comment (barbudor) [inline tasmota/include/tasmota_configurations_ESP32.h:692]: Do not enable your drivers by default
Also comment the increase in code size for ESP8266 (and memory too would be great)
Comment those lines so people are aware of what should be used

File: tasmota/tasmota_xsns_sensor/xsns_111_ens16x.ino
Comment (barbudor) [inline tasmota/tasmota_xsns_sensor/xsns_111_ens16x.ino:60]: Use dynamically allocated memory
This is using memory in any built your driver is included even if no-one is using it

**Review by barbudor — CHANGES_REQUESTED** [review]
Normally you should have done 2 PR, one for each sensor

ALso please advise code increase and memory increase for each of the drivers for both ESP8266 and ESP32
Thanks

File: tasmota/include/tasmota_configurations.h
Comment (chrfriese123) [reply tasmota/include/tasmota_configurations.h]: Disabled by default, memory information added

File: BUILDS.md
Comment (chrfriese123) [reply BUILDS.md]: Disabled by default, memory information added

File: tasmota/include/tasmota_configurations_ESP32.h
Comment (chrfriese123) [reply tasmota/include/tasmota_configurations_ESP32.h]: Disabled by default, memory information added

File: tasmota/tasmota_xsns_sensor/xsns_111_ens16x.ino
Comment (chrfriese123) [reply tasmota/tasmota_xsns_sensor/xsns_111_ens16x.ino:60]: Thanks for proposal. I'll try to implement later.

**Comment by chrfriese123** [comment]
Should I still split the pull request for both sensors?

**Comment by barbudor** [comment]
Hi @chrfriese123
Thanks
I'll let @arendst comment if he wants you to split

I find the code increase is quite high. Is the whole libraries needed for each sensors?
I've seen quite some large sensors libraries with code that is not used in general use case. But once a C/Cpp module is merge, all the code comes with it. There a few cases where it was most efficient to extract only the needed pieces.
I'll have further look tomorrow

Thanks

**Comment by Jason2866** [comment]
@chrfriese123 Is the ENS160 working with this driver? Are there development boards available for the ENS210? Where to get and what do the cost?

**Comment by chrfriese123** [comment]
> @chrfriese123 Is the ENS160 working with this driver? Are there development boards available for the ENS210? Where to get and what do the cost?

@Jason2866 the driver is working with ENS160 and ENS161. TVOC and eCO2 output is the same. Difference is only the air quality output and the low power option for ENS161.
Development boards for ENS16x are available from Adafruit and Sparkfun. The ENS210 from Sparkfun seems to be out of stock. The combined eval kit from Mouser is (unfortunately) bloody expensive.

**Comment by chrfriese123** [comment]
> Hi @chrfriese123 Thanks I'll let @arendst comment if he wants you to split
> 
> I find the code increase is quite high. Is the whole libraries needed for each sensors? I've seen quite some large sensors libraries with code that is not used in general use case. But once a C/Cpp module is merge, all the code comes with it. There a few cases where it was most efficient to extract only the needed pieces. I'll have further look tomorrow
> 
> Thanks

@barbudor, thanks for suggestion. Ideally I don’t want to diverge too much from the official driver. But I’ll give it a try to reduce to a minimum.

**Comment by barbudor** [comment]
We don't care about the official driver. What we care about is Flash and RAM usage in tasmota

I can confirm that there is a LOT a useless code in that library that you will never be used, starting with all the debug code full of Serial.print : 😱  This is totally forbidden in tasmota as the serial could be in use for something else.

I'm trying a first cut down

**Comment by barbudor** [comment]
Working just on the ENS16x while both are enabled, I get
```
Without 
RAM:   [=====     ]  51.3% (used 42000 bytes from 81920 bytes)
Flash: [======    ]  63.1% (used 646340 bytes from 1023984 bytes)

With both enabled
RAM:   [=====     ]  52.9% (used 43376 bytes from 81920 bytes)          + 1376
Flash: [======    ]  63.8% (used 653108 bytes from 1023984 bytes)       + 6768

Cut down on ENS16x
RAM:   [=====     ]  52.7% (used 43136 bytes from 81920 bytes)          + 1136 => gain 240 bytes
Flash: [======    ]  63.7% (used 652204 bytes from 1023984 bytes)       + 5864 => gain 904 bytes
```

So gaining 940 bytes on announced 3100 bytes cost (from your own comments) is not neglictable
I believe that reworking the I2C access could probably allow to reach the 1KB of gain

I also see a while loop with an delay(10) inside. That should be totally forbidden on a shared resource system like Tasmota.

I understand who you are and totally respect your will to provide an official library to your customers that support maximum of use cases. That work fine for someone who build a a product with a handfull of given sensors with a custom built firmware for that product.
But here we are trying to have a general purpose software that can handle as many features that can collaborate seamlessly together.

Let me know have a look to the ENS210

**Comment by barbudor** [comment]
I've seen ENS160 modules at Mouser from Sparkfun and DFrobot but none for ENS210
I've also seen combined modules with ENS160+ENS210 at AliExpress: originals or clones?

**Comment by chrfriese123** [comment]
The size cutdown is impressive. The serial.print is indeed a left over from the initial phase.
The while loop with delay is implemented to wait for new available data. If the read out cycle off longer than 1 sec, the register content can be read directly.
Also everything related to reading raw values is not required for everyday users and can be removed. The same is valid for defining the customer mode. These functions are not needed in a standard environment. 
For the ENS210 I don’t know the code so well but there might be options to reduce the code as well.
(Btw. this is a completely private activity from my side. I use Tasmota in my own devices to operate lights, plugs & covers and at the same time monitor the room air quality. Of course I have a sensor preference due to my professional backgrounds :-).)

**Comment by chrfriese123** [comment]
@barbudor I haven’t seen a combined eval kit at Aliexpress. Only one with a third party RHT sensor. 
The ones you find on Mouser from Sciosense are the official ones (white I2C to USB bridge + sensor board). But IMHO too expensive. Mainly because it’s not considered as end user product.
every other eval kit is a clone but not necessarily bad.

**Comment by barbudor** [comment]
When I was saying "clone" I wasn't talking about hte module/eval-kit. I don't consider Adafruit/Sparkfun/DFRobot to do clones as far as they use legit ScioScience parts.
But this one https://www.aliexpress.com/item/1005004399782807.html is advertised to use ENS160 and ENS210. Indeed there is a lot of ENS160 + AHT, so may be this one is also lying about the ENS210

**Comment by barbudor** [comment]
Lastest cuts on both libs is now
```
Final Cut
RAM:   [=====     ]  51.7% (used 42344 bytes from 81920 bytes)          +  344 => total gain 1032 bytes
Flash: [======    ]  63.5% (used 650188 bytes from 1023984 bytes)       + 3848 => total gain 2920 bytes
```

We are almost at 50% saving in flash and 75% saing in RAM

Please have a look at https://github.com/chrfriese123/Tasmota/pull/1
This is how I suggest to introduce 2 defines per lib which allows to cut down the code
You could leave them commented for in your official driver for non-breaking, and uncomment them when used in Tasmota
May be some customers of yours could also find valuable the savings

So why is that so huge and also impact the RAM ?
You have a lot of Serial.print() for debug with strings
That is using quite some flash for the code, but also for the strings.
But why over 1KB of saving in RAM ?
When you have a code like:
```
Serial.println("Hello world!");
```
The string is saved in flash but at boot time it is copied to the RAM because the ESP are an modified Harward architecture and you need different instructions to read RAM or Flash. So to simplify the code, all constants are loaded from Flash to RAM.
But in Tasmota we want to save that precious RAM.
The compiler offer options such as the `PROGMEM` qualifier or `PSTR(".....")` so that constants including strings remains in Flash. In that case, the code using them must be able to distinguish between Flash and RAM addresse and uses the appropriate instructions.


I'll have a look at the Tasmora code now

**Comment by barbudor** [comment]
2 hours later:
```
RAM:   [=====     ]  51.3% (used 42052 bytes from 81920 bytes)          +   52 => final gain 1324 bytes
Flash: [======    ]  63.5% (used 649916 bytes from 1023984 bytes)       + 3576 => final gain 3182 bytes
```

I optimized the Tasmota code:
- moving to dynamically allocated memory saved a lot
- de-duplicating the code with indexed data

As you can see, spending a little bit of time helps a lot and leave more room for other stuff.
I don't pretend that all drivers are at that level, and I don't pretend it wouldn't possible to go further on this one.

STILL: I am unable to test the changes*, especially the latests ones which may impact more the logic.
So please test what I have pushed to your fork or available from my branch here: https://github.com/barbudor/Tasmota/tree/barbudor_optimize_take1

I hope you will not take offense for my razor cut and will accept the changes

Thanks

image

_* if ScioSense has some budget for supporting open source projects with a little eval board ... 👀_

**Comment by chrfriese123** [comment]
Thanks, for the cut down and no offence taken. I'll compile the code and run a physical test. 

_* send you a mail_

**Comment by Jason2866** [comment]
@chrfriese123 Will you implement the suggestions?

**Comment by barbudor** [comment]
@Jason2866 I have already done the changes and will test this week end as I have received samples

**Comment by Jason2866** [comment]
@barbudor No hurry, just asking if there is activity done or planned.

**Comment by barbudor** [comment]
Working so far except for a bug ENS16X's WebDisplay that is crashing the ESP
I've pushed the change to @chrfriese123 repo. As soon as he will merge we are good to go
https://github.com/chrfriese123/Tasmota/pull/2/commits/adc990215a07492fa58d57ed6eccb42a7ff8b4f7

**Comment by barbudor** [comment]
Arg, detection of alternate addresses still have a problem

**Comment by barbudor** [comment]
All good now

image

```
14:33:59.211 RSL: SENSOR = {"Time":"2023-09-23T14:33:59","EN16X_1":{"AQIS":8,"eCO2":400,"TVOC":0},"EN16X_2":{"AQIS":86,"eCO2":590,"TVOC":61},"EN210":{"Temperature":23.4,"Humidity":54.7,"DewPoint":13.7},"TempUnit":"C"}
```

ENS16X: +1984 Flash bytes, +24 RAM bytes
ENS210: +1740 Flash bytes, +24 RAM bytes
Strangely when both are enabled, the increase is lower than the sum of both separately
No idea where that additional optimization comes from as there is no common variables or code between both
ENS16X + ENS210: +3527 Flash bytes (vs 3724 expected), +28 RAM bytes (vs 56 expected)

**Comment by barbudor** [comment]
@Jason2866 Still need @chrfriese123 to merge my changes in his repo to see all changes here

**Comment by chrfriese123** [comment]
@Jason2866, @barbudor the last changes have been merged in my repo. Thanks for the great support in fixing the code.

**Pull request merged**
